patches/server/0928-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,1156 @@
++        } else {
++            throw new IllegalArgumentException("Unknown command node passed.");
++        }
++
++        for (CommandNode<CommandSourceStack> child : nms.getChildren()) {
++            rebuiltNode.addChild(this.unwrapWrappedArguments(child));
++        }
++        this.cachedUnwrappedArguments.put(nms, rebuiltNode);
++
++        return rebuiltNode;
++    }
++
++    private CommandNode<CommandSourceStack> createWrappedArgument(ArgumentCommandNode node, WrapperArgumentType wrapperArgumentType) {
++        CommandNode redirectNode = node.getRedirect() == null ? null : this.unwrapWrappedArguments(node.getRedirect());
++        SuggestionProvider suggestionProvider;
++        // If there is already a custom suggestion provider, ignore.
++        if (node.getCustomSuggestions() != null) {
++            suggestionProvider = node.getCustomSuggestions();
++        } else {
++            // If however we want to handle suggestions ourselves, reference the wrapper.
++            suggestionProvider = wrapperArgumentType.handleSuggestions() ? wrapperArgumentType::listSuggestions : null;
++        }
++
++        return new WrappedArgumentCommandNode<>(node.getName(), wrapperArgumentType, wrapperArgumentType.getWrapped(), node.getCommand(), node.getRequirement(), redirectNode, node.getRedirectModifier(), node.isFork(), suggestionProvider);
++    }
++
++    private CommandNode<CommandSourceStack> recreateLiteral(LiteralCommandNode<CommandSourceStack> node) {
++        CommandNode redirectNode = node.getRedirect() == null ? null : this.unwrapWrappedArguments(node.getRedirect());
++
++        return new LiteralCommandNode<>(node.getLiteral(), node.getCommand(), node.getRequirement(), redirectNode, node.getRedirectModifier(), node.isFork());
++    }
++
++    @Override
++    public void addChild(CommandNode<CommandSourceStack> node) {
++        this.apiChildren.add(node);
++        CommandNode convertedNode = this.unwrapWrappedArguments(node);
++        this.nmsSourceStack.getRoot().addChild(convertedNode);
++    }
++
++    @Override
++    public Collection<CommandNode<CommandSourceStack>> getChildren() {
++        return this.apiChildren;
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerImpl.java b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerImpl.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..48e55b763b0c92ada24e624b2fe0eb50d187229d
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerImpl.java
+@@ -0,0 +1,20 @@
++package io.papermc.paper.command.brigadier;
++
++import com.mojang.brigadier.Message;
++import io.papermc.paper.adventure.PaperAdventure;
++import net.kyori.adventure.text.Component;
++import net.minecraft.network.chat.ComponentUtils;
++import org.jetbrains.annotations.NotNull;
++
++public class MessageComponentSerializerImpl implements MessageComponentSerializer {
```

Comment:
I think these classes can be non-public (+final)? I think that works with the ServiceLoader